function argument safety check silently optimized out in release build by clang

### Description

With `clang 21.1.8` on `macOS`, or `clang-cl.exe 21.1.8` on `Windows` the release build of unit test `test-class_parser_cpp11` (`tests/src/unit-class_parser.cpp`) crashes with an unhandled exception (`SIGSEGV` on macOS, `Access violation` on Windows).

The reason why it crashes is that in release build the compiler optimizes out the safety check here:
https://github.com/nlohmann/json/blob/515d994acbac04e0efa41db21e0d16591f6f97c9/include/nlohmann/json.hpp#L4132

Normally, there would be a warning emitted by the compiler:
```shell
/Users/risa/Dev/LLVM/bin/clang++ -DDOCTEST_CONFIG_SUPER_FAST_ASSERTS -DJSON_TEST_KEEP_MACROS -DJSON_TEST_USING_MULTIPLE_HEADERS=1 -I/Users/risa/Work_OSS/_risa2000/nlohmann_json/tests/thirdparty/doctest -I/Users/risa/Work_OSS/_risa2000/nlohmann_json/tests/thirdparty/fifo_map -I/Users/risa/Work_OSS/_risa2000/nlohmann_json/out/build/arm64_RelWithDebInfo/include -I/Users/risa/Work_OSS/_risa2000/nlohmann_json/include -O2 -g -DNDEBUG -std=gnu++11 -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -Wno-deprecated -Wno-float-equal -o tests/CMakeFiles/test-class_parser_cpp11.dir/src/unit-class_parser.cpp.o -c /Users/risa/Work_OSS/_risa2000/nlohmann_json/tests/src/unit-class_parser.cpp
In file included from /Users/risa/Work_OSS/_risa2000/nlohmann_json/tests/src/unit-class_parser.cpp:12:
/Users/risa/Work_OSS/_risa2000/nlohmann_json/include/nlohmann/json.hpp:4132:13: warning: comparison of nonnull parameter 'sax' equal to a null pointer is 'false' on first
      encounter [-Wtautological-pointer-compare]
 4132 |         if (sax == nullptr)
      |             ^~~    ~~~~~~~
/Users/risa/Work_OSS/_risa2000/nlohmann_json/tests/src/unit-class_parser.cpp:224:11: note: in instantiation of function template specialization
      'nlohmann::basic_json<>::sax_parse<const std::string &, nlohmann::detail::json_sax_dom_parser<nlohmann::basic_json<>, nlohmann::detail::iterator_input_adapter<std::__wrap_iter<const char
      *>>>>' requested here
  224 |     json::sax_parse(s, &sdp);
      |           ^
/Users/risa/Work_OSS/_risa2000/nlohmann_json/include/nlohmann/json.hpp:4118:5: note: declared 'nonnull' here
 4118 |     JSON_HEDLEY_NON_NULL(2)
      |     ^
/Users/risa/Work_OSS/_risa2000/nlohmann_json/include/nlohmann/thirdparty/hedley/hedley.hpp:1287:54: note: expanded from macro 'JSON_HEDLEY_NON_NULL'
 1287 |     #define JSON_HEDLEY_NON_NULL(...) __attribute__((__nonnull__(__VA_ARGS__)))
      |                                                      ^
In file included from /Users/risa/Work_OSS/_risa2000/nlohmann_json/tests/src/unit-class_parser.cpp:12:
/Users/risa/Work_OSS/_risa2000/nlohmann_json/include/nlohmann/json.hpp:4132:13: warning: comparison of nonnull parameter 'sax' equal to a null pointer is 'false' on first
      encounter [-Wtautological-pointer-compare]
 4132 |         if (sax == nullptr)
      |             ^~~    ~~~~~~~
/Users/risa/Work_OSS/_risa2000/nlohmann_json/tests/src/unit-class_parser.cpp:251:25: note: in instantiation of function template specialization
      'nlohmann::basic_json<>::sax_parse<const std::string &, (anonymous namespace)::SaxEventLogger>' requested here
  251 |     CHECK_NOTHROW(json::sax_parse(s, &el, json::input_format_t::json, false));
      |                         ^
/Users/risa/Work_OSS/_risa2000/nlohmann_json/tests/thirdparty/doctest/doctest.h:2976:50: note: expanded from macro 'CHECK_NOTHROW'
 2976 | #define CHECK_NOTHROW(...) DOCTEST_CHECK_NOTHROW(__VA_ARGS__)
      |                                                  ^
/Users/risa/Work_OSS/_risa2000/nlohmann_json/tests/thirdparty/doctest/doctest.h:2544:77: note: expanded from macro 'DOCTEST_CHECK_NOTHROW'
 2544 | #define DOCTEST_CHECK_NOTHROW(...) DOCTEST_ASSERT_NOTHROW(DT_CHECK_NOTHROW, __VA_ARGS__)
      |                                                                             ^
/Users/risa/Work_OSS/_risa2000/nlohmann_json/tests/thirdparty/doctest/doctest.h:2521:34: note: expanded from macro 'DOCTEST_ASSERT_NOTHROW'
 2521 |             DOCTEST_CAST_TO_VOID(__VA_ARGS__)                                                      \
      |                                  ^
/Users/risa/Work_OSS/_risa2000/nlohmann_json/tests/thirdparty/doctest/doctest.h:2154:23: note: expanded from macro 'DOCTEST_CAST_TO_VOID'
 2154 |     static_cast<void>(__VA_ARGS__);                                                                \
      |                       ^
/Users/risa/Work_OSS/_risa2000/nlohmann_json/include/nlohmann/json.hpp:4118:5: note: declared 'nonnull' here
 4118 |     JSON_HEDLEY_NON_NULL(2)
      |     ^
/Users/risa/Work_OSS/_risa2000/nlohmann_json/include/nlohmann/thirdparty/hedley/hedley.hpp:1287:54: note: expanded from macro 'JSON_HEDLEY_NON_NULL'
 1287 |     #define JSON_HEDLEY_NON_NULL(...) __attribute__((__nonnull__(__VA_ARGS__)))
      |                                                      ^
```
but the warning `-Wtautological-pointer-compare` is silenced in the library code, and it seems this warning is actually trying to warn about this.

The debug build does not exhibit this problem as there the check is compiled fine.

I am not sure if this is a bug in compiler (though it seems the behavior is there since quite some time) which is too ambitious, but there is a compiler flag, which disables this optimization even in the release build `-fno-delete-null-pointer-checks` (supposedly available from `clang 7`), but the problem is, the library client must add this flag manually when using the library code.

There seems to be the similar warning (treatment) for `gcc`, but I do not have a `gcc` environment to test, if this warning is harmless, or produces the same problem.

I guess the best solution would be rewriting the code in a way it does not fall for this optimization, but so far I did not find how or if possible at all (I am not that much familiar with `clang`).

It is also strange that it does not get caught by CI/CD tests.

### Reproduction steps

Run the release build of `test-class_parser_cpp11` unit test, built with the recent version (`21.1.8`) of `clang` (on macOS), or `clang-cl.exe` (on Windows).

### Expected vs. actual results

Should not crash.

### Minimal code example

```Shell

```

### Error messages

```Shell
[doctest] doctest version is "2.4.12"
[doctest] run with "--help" for options
===============================================================================
/Users/risa/Work_OSS/_risa2000/nlohmann_json/tests/src/unit-class_parser.cpp:343:
TEST CASE:  parser class
  improve test coverage
  SAX parser
  null sax handler

/Users/risa/Work_OSS/_risa2000/nlohmann_json/tests/src/unit-class_parser.cpp:343: FATAL ERROR: test case CRASHED: SIGSEGV - Segmentation violation signal
```

### Compiler and operating system

`clang` (macOS), `clang-cl.exe` (Windows), version `21.1.8`.

### Library version

develop

### Validation

- [x] The bug also occurs if the latest version from the [`develop`](https://github.com/nlohmann/json/tree/develop) branch is used.
- [ ] I can successfully [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests).

--- a/docs/mkdocs/docs/home/exceptions.md
+++ b/docs/mkdocs/docs/home/exceptions.md
@@ -900,13 +900,3 @@ A JSON Patch operation 'test' failed. The unsuccessful operation is also printed
     ```
     [json.exception.other_error.501] unsuccessful: {"op":"test","path":"/baz","value":"bar"}
     ```
-
-### json.exception.other_error.502
-
-This exception is thrown when a null pointer is passed as SAX event listener to the [sax_parse](../api/basic_json/sax_parse.md) function.
-
-!!! failure "Example message"
-
-    ```
-    [json.exception.other_error.502] SAX handler must not be null
-    ```

--- a/include/nlohmann/json.hpp
+++ b/include/nlohmann/json.hpp
@@ -4122,22 +4122,6 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
                           const bool ignore_comments = false,
                           const bool ignore_trailing_commas = false)
     {
-#if defined(__clang__)
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wtautological-pointer-compare"
-#elif defined(__GNUC__)
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wnonnull-compare"
-#endif
-        if (sax == nullptr)
-        {
-            JSON_THROW(other_error::create(502, "SAX handler must not be null", nullptr));
-        }
-#if defined(__clang__)
-#pragma clang diagnostic pop
-#elif defined(__GNUC__)
-#pragma GCC diagnostic pop
-#endif
         auto ia = detail::input_adapter(std::forward<InputType>(i));
         return format == input_format_t::json
                ? parser(std::move(ia), nullptr, true, ignore_comments, ignore_trailing_commas).sax_parse(sax, strict)
@@ -4154,22 +4138,6 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
                           const bool ignore_comments = false,
                           const bool ignore_trailing_commas = false)
     {
-#if defined(__clang__)
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wtautological-pointer-compare"
-#elif defined(__GNUC__)
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wnonnull-compare"
-#endif
-        if (sax == nullptr)
-        {
-            JSON_THROW(other_error::create(502, "SAX handler must not be null", nullptr));
-        }
-#if defined(__clang__)
-#pragma clang diagnostic pop
-#elif defined(__GNUC__)
-#pragma GCC diagnostic pop
-#endif
         auto ia = detail::input_adapter(std::move(first), std::move(last));
         return format == input_format_t::json
                ? parser(std::move(ia), nullptr, true, ignore_comments, ignore_trailing_commas).sax_parse(sax, strict)
@@ -4190,22 +4158,6 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
                           const bool ignore_comments = false,
                           const bool ignore_trailing_commas = false)
     {
-#if defined(__clang__)
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wtautological-pointer-compare"
-#elif defined(__GNUC__)
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wnonnull-compare"
-#endif
-        if (sax == nullptr)
-        {
-            JSON_THROW(other_error::create(502, "SAX handler must not be null", nullptr));
-        }
-#if defined(__clang__)
-#pragma clang diagnostic pop
-#elif defined(__GNUC__)
-#pragma GCC diagnostic pop
-#endif
         auto ia = i.get();
         return format == input_format_t::json
                // NOLINTNEXTLINE(hicpp-move-const-arg,performance-move-const-arg)

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -24359,22 +24359,6 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
                           const bool ignore_comments = false,
                           const bool ignore_trailing_commas = false)
     {
-#if defined(__clang__)
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wtautological-pointer-compare"
-#elif defined(__GNUC__)
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wnonnull-compare"
-#endif
-        if (sax == nullptr)
-        {
-            JSON_THROW(other_error::create(502, "SAX handler must not be null", nullptr));
-        }
-#if defined(__clang__)
-#pragma clang diagnostic pop
-#elif defined(__GNUC__)
-#pragma GCC diagnostic pop
-#endif
         auto ia = detail::input_adapter(std::forward<InputType>(i));
         return format == input_format_t::json
                ? parser(std::move(ia), nullptr, true, ignore_comments, ignore_trailing_commas).sax_parse(sax, strict)
@@ -24391,22 +24375,6 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
                           const bool ignore_comments = false,
                           const bool ignore_trailing_commas = false)
     {
-#if defined(__clang__)
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wtautological-pointer-compare"
-#elif defined(__GNUC__)
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wnonnull-compare"
-#endif
-        if (sax == nullptr)
-        {
-            JSON_THROW(other_error::create(502, "SAX handler must not be null", nullptr));
-        }
-#if defined(__clang__)
-#pragma clang diagnostic pop
-#elif defined(__GNUC__)
-#pragma GCC diagnostic pop
-#endif
         auto ia = detail::input_adapter(std::move(first), std::move(last));
         return format == input_format_t::json
                ? parser(std::move(ia), nullptr, true, ignore_comments, ignore_trailing_commas).sax_parse(sax, strict)
@@ -24427,22 +24395,6 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
                           const bool ignore_comments = false,
                           const bool ignore_trailing_commas = false)
     {
-#if defined(__clang__)
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wtautological-pointer-compare"
-#elif defined(__GNUC__)
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wnonnull-compare"
-#endif
-        if (sax == nullptr)
-        {
-            JSON_THROW(other_error::create(502, "SAX handler must not be null", nullptr));
-        }
-#if defined(__clang__)
-#pragma clang diagnostic pop
-#elif defined(__GNUC__)
-#pragma GCC diagnostic pop
-#endif
         auto ia = i.get();
         return format == input_format_t::json
                // NOLINTNEXTLINE(hicpp-move-const-arg,performance-move-const-arg)

--- a/tests/src/unit-class_parser.cpp
+++ b/tests/src/unit-class_parser.cpp
@@ -1644,33 +1644,6 @@ TEST_CASE("parser class")
 
         SECTION("SAX parser")
         {
-            SECTION("null sax handler")
-            {
-# if defined(__has_feature)
-#if !__has_feature(undefined_behavior_sanitizer)
-                const std::string s = "some_string";
-                SaxCountdown* p = nullptr;
-                CHECK_THROWS_WITH_AS(json::sax_parse(s, p), "[json.exception.other_error.502] SAX handler must not be null", json::other_error&); // NOLINT(clang-analyzer-core.NonNullParamChecker)
-                CHECK_THROWS_WITH_AS(json::sax_parse(s.begin(), s.end(), p), "[json.exception.other_error.502] SAX handler must not be null", json::other_error&); // NOLINT(clang-analyzer-core.NonNullParamChecker)
-                CHECK_THROWS_WITH_AS(json::sax_parse(nlohmann::detail::span_input_adapter(s.c_str(), s.size()), p), "[json.exception.other_error.502] SAX handler must not be null", json::other_error&); // NOLINT(clang-analyzer-core.NonNullParamChecker)
-#endif
-#else
-                const std::string s = "some_string";
-                SaxCountdown* p = nullptr;
-                CHECK_THROWS_WITH_AS(json::sax_parse(s, p), "[json.exception.other_error.502] SAX handler must not be null", json::other_error&); // NOLINT(clang-analyzer-core.NonNullParamChecker)
-                CHECK_THROWS_WITH_AS(json::sax_parse(s.begin(), s.end(), p), "[json.exception.other_error.502] SAX handler must not be null", json::other_error&); // NOLINT(clang-analyzer-core.NonNullParamChecker)
-                CHECK_THROWS_WITH_AS(json::sax_parse(nlohmann::detail::span_input_adapter(s.c_str(), s.size()), p), "[json.exception.other_error.502] SAX handler must not be null", json::other_error&); // NOLINT(clang-analyzer-core.NonNullParamChecker)
-#endif
-            }
-
-            SECTION("valid sax handler")
-            {
-                const std::string str = "some_string";
-                SaxCountdown s(1);
-                CHECK(json::sax_parse(str, &s) == false);
-                CHECK(json::sax_parse(nlohmann::detail::span_input_adapter(str.c_str(), str.size()), &s) == false);
-            }
-
             SECTION("} without value")
             {
                 SaxCountdown s(1);

